content/2023-05-03-beware-of-unified-builds.rst

```diff
@@ -0,0 +1,317 @@
+How Jumbo builds lurked into the Firefox Build System
+#####################################################
+
+:date: 2023-05-03
+:category: mozilla
+:lang: en
+:authors: Serge Guelton
+:summary: Jumbo builds make C++ code compilation faster, but what happens when
+          you require them?
+
+
+Jumbo Build
+===========
+
+Jumbo builds, a.k.a. unified build or unity builds, is a technique that consists in
+concatenating several C or C++ files in one before compiling. This generally
+leads to faster compilation time because it aggregates the cost of parsing the
+same headers over and over.
+
+It is one of the many approach one can use to reduce C++ software compilation
+time, alongside precompiled headers and C++20 modules.
+
+It is not an obscure technique: it's officially supported by CMake through
+``CMAKE_UNITY_BUILD`` (see
+https://cmake.org/cmake/help/latest/variable/CMAKE_UNITY_BUILD.html)
+
+At some point, chromium has been using it:
+https://chromium.googlesource.com/chromium/src.git/+/65.0.3283.0/docs/jumbo.md,
+even if they get rid of it afterward
+https://groups.google.com/a/chromium.org/g/chromium-dev/c/DP9TQszzQLI
+
+It is also supported by the internal build system used at Mozilla. And the
+speedup is there, a unified build (the default) runs twice as fast as an hybrid
+build (``--disable-unified-build``) on my setup.
+
+Wait, did I write *hybrid build* and not *regular build*? As it turns out, under
+``--disable-unified-build`` some parts of Firefox are still built in unified
+mode, because they **require it**, probably for historical reason.
+
+It's great to be able to do a unified build. It's not great to have a codebase
+that does not compile unless you have a unity build: static analyzers are not
+used to work on non-self contained sources (see https://github.com/clangd/clangd/issues/45), unity build implies a slight overhead during
+incremental builds. What makes it worse is that developer start to rely on
+unified build and get lazy in the way they develop.
```

Comment:
We're not lazy, we simply have no straightforward way to know there's an error, short of doing a regular build, which we don't want to do, because of the reasons we're doing unified builds.

Reply:
Here it's important to note that it wasn't really possible to do a non-unified build, because the next paragraph talks about fixing things that prevented it -- this fact hasn't been stated yet despite being the entire point of the article.